pkg/template/controller/templateinstance_controller.go

```diff
@@ -132,38 +140,131 @@ func (c *TemplateInstanceController) sync(key string) error {
 		return err
 	}
 
-	instantiateErr := c.instantiate(templateInstance)
-	if instantiateErr != nil {
-		glog.V(4).Infof("TemplateInstance controller: instantiate %s returned %v", key, instantiateErr)
+	if len(templateInstance.Status.Objects) != len(templateInstance.Spec.Template.Objects) {
+		err = c.instantiate(templateInstance)
+		if err != nil {
+			glog.V(4).Infof("TemplateInstance controller: instantiate %s returned %v", key, err)
+
+			templateInstance.SetCondition(templateapi.TemplateInstanceCondition{
+				Type:    templateapi.TemplateInstanceInstantiateFailure,
+				Status:  kapi.ConditionTrue,
+				Reason:  "Failed",
+				Message: err.Error(),
+			})
+		}
 	}
 
-	templateInstance.Status.Conditions = templateapi.FilterTemplateInstanceCondition(templateInstance.Status.Conditions, templateapi.TemplateInstanceReady)
-	templateInstance.Status.Conditions = templateapi.FilterTemplateInstanceCondition(templateInstance.Status.Conditions, templateapi.TemplateInstanceInstantiateFailure)
-
-	if instantiateErr == nil {
-		templateInstance.Status.Conditions = append(templateInstance.Status.Conditions, templateapi.TemplateInstanceCondition{
-			Type:               templateapi.TemplateInstanceReady,
-			Status:             kapi.ConditionTrue,
-			LastTransitionTime: metav1.Now(),
-			Reason:             "Created",
-		})
-
-	} else {
-		templateInstance.Status.Conditions = append(templateInstance.Status.Conditions, templateapi.TemplateInstanceCondition{
-			Type:               templateapi.TemplateInstanceInstantiateFailure,
-			Status:             kapi.ConditionTrue,
-			LastTransitionTime: metav1.Now(),
-			Reason:             "Failed",
-			Message:            instantiateErr.Error(),
-		})
+	if !templateInstance.HasCondition(templateapi.TemplateInstanceInstantiateFailure, kapi.ConditionTrue) {
+		ready, err := c.checkReadiness(templateInstance)
+		if err != nil {
+			glog.V(4).Infof("TemplateInstance controller: checkReadiness %s returned %v", key, err)
+
+			templateInstance.SetCondition(templateapi.TemplateInstanceCondition{
+				Type:    templateapi.TemplateInstanceInstantiateFailure,
+				Status:  kapi.ConditionTrue,
+				Reason:  "Failed",
+				Message: err.Error(),
+			})
+			templateInstance.SetCondition(templateapi.TemplateInstanceCondition{
+				Type:   templateapi.TemplateInstanceReady,
+				Status: kapi.ConditionFalse,
```

Comment:
should this have a reason/message?  maybe redundant given that the failed condition is asserted, but it seems like it should be here for completeness (reason waiting==false: instantiate has failed)